### PR TITLE
[new featues] log & manager

### PR DESCRIPTION
**Pull request by ZJUMark**
[features] log with level&color/auto multicast manager/auto add info of pub&sub to manager and send multicast

## Proposed changes

* add log with levels[log&status&info&warning&error] and different colors
* add manager add auto multicast info of pubers&subers with default 3hz

## Types of changes

What types of changes does your code introduce to Appium?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update (if none of the other choices apply)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] I have read the [CONTRIBUTING](https://github.com/ZJUNlict/zos/blob/master/CONTRIBUTING.md) doc
- [x] Lint and unit tests pass locally with my changes
- [ ] I have added annotation in every module or function of my work
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation like README.md (if appropriate)
- [x] Any dependent changes have been merged and published in downstream modules

## Further comments

future works:
* pubers&subers with socket support
* GUI with manager
  * visualizer for pubers&subers&links
  * handle for adding links between two different nodes (auto create socket connecting)
  * handle other zos process in LAN not only in localhost
* node init function for manger and other modules
* add more support for callback functions in socket and subscribers using template instead of specific type

### CODE REVIEW

File: project/core/include/zos/log.h [hunk 1]
```diff
@@ -12,27 +13,72 @@ namespace __impl{
 typedef Singleton<std::mutex> LogMutex;
 template<typename... Ts>
 struct __Log{
-    __Log(std::string_view s,Ts&&... ts,const std::source_location& lo=std::source_location::current()){
+    __Log(fmt::text_style style,std::string_view s,const Ts&... ts,const std::source_location& lo=std::source_location::current()){
         std::string total_filename{lo.file_name()};
         auto found = total_filename.find_last_of('/');
```
Comment (tyGavinZJU) [inline project/core/include/zos/log.h:18]: 如果windows的话，应该写
```c++
.find_last_of('/\\')
```

File: project/core/include/zos/utils/rate.h [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+#include <thread>
+namespace zos{
+class Rate{
+public:
+    Rate(const int frq){
+        _last_t = std::chrono::steady_clock::now();
+        _step = std::chrono::nanoseconds(1000000000)/frq;
+    }
+    void sleep(){
+        std::this_thread::sleep_until(_last_t+_step);
```
Comment (tyGavinZJU) [inline project/core/include/zos/utils/rate.h:11]: `_last_t` is not the exactly now() when `Rate` object call `sleep`

``` c++
auto awake_time(auto step) {
    return now() + step;
}

void sleep(){
    std::this_thread::sleep_until(awake_time(_step))
}
```

File: project/core/include/zos/utils/singleton.h [hunk 3]
```diff
@@ -1,24 +1,33 @@
...
+    static T* _(Ts... args){ return instance(args...); }
 private:
     Singleton() = default;
-    ~Singleton() = default;
+    ~Singleton(){
+        delete inst;
+    };
     Singleton( const Singleton& ) = delete;
     Singleton& operator=( const Singleton& ) = delete;
+    static inline T* inst = nullptr;
```
Comment (tyGavinZJU) [inline project/core/include/zos/utils/singleton.h:23]: https://codereview.stackexchange.com/questions/173929/modern-c-singleton-template#comment330047_173935

这个讲了如果单例用 static inline来做可能会导致并发构造的时候会失败，因为无法确保进程安全。这块太细节了你研究一下吧 @ZJUMark

File: project/core/include/zos/protocol/zos.proto [hunk 4]
```diff
@@ -1,15 +1,32 @@
...
+    repeated string sub_names = 5;
 }
 message EndPoint{
     string address = 1;
     int32 port = 2;
 }
 message State{
-    EndPoint ep = 1;
-    repeated Msg msgs = 2;
+    enum ProcessState {
```
Comment (tyGavinZJU) [inline project/core/include/zos/protocol/zos.proto:21]: 感觉proto文件写文档优先级最高，要不然manager很难读懂啊

File: project/core/include/zos/core.h [hunk 5]
```diff
@@ -27,7 +28,7 @@ class Pool:public Singleton<Pool>,public ThreadPool{
 };
 class SubscriberBase{
 public:
-    SubscriberBase(const std::string& msg,const __callback_type& f = {}):_data(nullptr),_msg(msg){
+    SubscriberBase(const std::string& name,const std::string& msg_name,const __callback_type& f = {}):_data(nullptr),_name(name),_msg_name(msg_name){
```
Comment (tyGavinZJU) [inline project/core/include/zos/core.h:31]: subscriber/publisher 功能上的文档支持

File: project/core/include/zos/manager.h [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+#include "zos/config.h"
+#include "zos/utils/rate.h"
+#include "zos/log.h"
+#include "zos/socket.h"
+#include "zos.pb.h"
+#include "zos/posix.h"
+namespace zos{
+// class SPublisher;
+// class SSubscriber;
+class manager:public Singleton<manager>{
```
Comment (tyGavinZJU) [inline project/core/include/zos/manager.h:13]: manager流程的文档支持

File: project/core/include/zos/manager.h [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+public:
+    manager(const char* _name="",const int rate=3,const char* _if=default_interface,const char* _address=default_mc_address,const int _port=default_port):_pid(getpid()),_name(_name),_endpoint(udp::address::from_string(_address),_port),_interface(_if),_mc_address(_address),_port(_port),_rate(Rate(rate)){
+        zos::warning("inside manager address : {},name-{}\n",fmt::ptr(this),_name);
+        if(this->_name.empty()){
+            zos::warning("zos manager name is empty,maybe you need to set one name on ctor.\n");
+        }
+        _write.set_interface(zos::udp::address::from_string(_interface));
+        _share_read.join_multicast(udp::address::from_string(_mc_address),udp::address::from_string(_interface));
+        _share_read.bind(udp::endpoint(udp::address_v4::any(),_port),std::bind(&manager::_cb,this,std::placeholders::_1,std::placeholders::_2));
+        _state_send = std::thread([this]{
```
Comment (tyGavinZJU) [inline project/core/include/zos/manager.h:26]: annotation

File: project/core/include/zos/log.h (hunk 1 above)
Comment (ZJUMark) [inline project/core/include/zos/log.h:18]: https://cplusplus.com/reference/string/string/find_last_of/
str.find_last_of("/\\");

File: project/core/include/zos/manager.h [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+    const int _port;
+    const udp::endpoint _endpoint;
+    const int _pid;
+
+    udp::socket _share_read;
+    udp::socket _unique_read;
+    udp::socket _write;
+    Rate _rate;
+
+    std::thread _asio_context;
```
Comment (tyGavinZJU) [inline project/core/include/zos/manager.h:131]: annotation

**Pull request merged**